services/src/handlers/datasets.rs

```diff
@@ -97,8 +139,189 @@ fn adjust_user_path_to_upload_path(meta: &mut MetaDataDefinition, upload: &Uploa
     Ok(())
 }
 
+pub(crate) fn auto_create_dataset_handler<C: Context>(
+    ctx: C,
+) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+    warp::path!("dataset" / "auto")
+        .and(warp::post())
+        .and(authenticate(ctx.clone()))
+        .and(warp::any().map(move || ctx.clone()))
+        .and(warp::body::json())
+        .and_then(auto_create_dataset)
+}
+
+// TODO: move into handler once async closures are available?
+async fn auto_create_dataset<C: Context>(
+    session: Session,
+    ctx: C,
+    create: AutoCreateDataSet,
+) -> Result<impl warp::Reply, warp::Rejection> {
+    let upload = ctx
+        .data_set_db_ref()
+        .await
+        .get_upload(session.user.id, create.upload)
+        .await?;
+
+    let create = create.validated()?.user_input;
+
+    let main_file_path = upload.id.root_path()?.join(&create.main_file);
+    let meta_data = auto_detect_dataset(&main_file_path)?;
+
+    let properties = AddDataSet {
+        id: None,
+        name: create.dataset_name,
+        description: create.dataset_description,
+        source_operator: "OgrSource".to_owned(),
+    };
+
+    let mut db = ctx.data_set_db_ref_mut().await;
+    let meta_data = db.wrap_meta_data(meta_data);
+    let id = db
+        .add_data_set(session.user.id, properties.validated()?, meta_data)
+        .await?;
+
+    Ok(warp::reply::json(&IdResponse::from(id)))
+}
+
+fn auto_detect_dataset(main_file_path: &Path) -> Result<MetaDataDefinition> {
+    let mut dataset = Dataset::open(&main_file_path).context(error::Gdal)?;
+    let layer = {
+        if let Ok(layer) = dataset.layer(0) {
+            layer
+        } else {
+            // TODO: handle Raster datasets as well
+            return Err(crate::error::Error::DataSetHasNoAutoImportableLayer);
+        }
+    };
+    let vector_type = detect_vector_type(&layer)?;
+    let spatial_reference: SpatialReference = layer
+        .spatial_ref()
+        .context(error::Gdal)?
+        .try_into()
+        .context(error::DataType)?;
+    let columns_map = detect_columns(&layer);
+    let columns_vecs = column_map_to_column_vecs(&columns_map);
+
+    Ok(MetaDataDefinition::OgrMetaData(StaticMetaData {
+        loading_info: OgrSourceDataset {
+            file_name: main_file_path.into(),
+            layer_name: layer.name(),
+            data_type: Some(vector_type),
+            time: OgrSourceDatasetTimeType::None, // TODO: auto detect time type and corresponding columns
+            columns: Some(OgrSourceColumnSpec {
+                x: "".to_owned(), // TODO: for csv-files: try to find wkt/xy columns
+                y: None,
```

Comment:
Does Ogr output geometry columns in general?

Reply:
Maybe you can use the information in https://docs.rs/gdal/0.7.2/gdal/vector/struct.Defn.html#method.geom_fields to get the columns.

Reply:
73b4e8430add8c17b2e5670d7a011c6fbc3fa5e4